crates/bevy_ecs/src/resource.rs

```diff
@@ -72,4 +86,103 @@ pub use bevy_ecs_macros::Resource;
     label = "invalid `Resource`",
     note = "consider annotating `{Self}` with `#[derive(Resource)]`"
 )]
-pub trait Resource: Send + Sync + 'static {}
+pub trait Resource: Send + Sync + 'static + MapEntities {}
+
+/// A cache that links each `ComponentId` from a resource to the corresponding entity.
+#[derive(Default)]
+pub struct ResourceCache(SyncUnsafeCell<SparseSet<ComponentId, Entity>>);
+
+impl ResourceCache {
+    fn inner(&self) -> &SparseSet<ComponentId, Entity> {
+        // SAFETY: pointer was just created, so it's convertible
+        unsafe { &*self.0.get() }
+    }
+
+    fn inner_mut(&mut self) -> &mut SparseSet<ComponentId, Entity> {
+        self.0.get_mut()
+    }
+
+    pub(crate) fn get(&self, id: ComponentId) -> Option<&Entity> {
+        self.inner().get(id)
+    }
+
+    pub(crate) fn contains(&self, id: ComponentId) -> bool {
+        self.inner().contains(id)
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.inner().len()
+    }
+
+    pub(crate) fn clear(&mut self) {
+        self.inner_mut().clear();
+    }
+
+    pub(crate) fn entities(&self) -> impl Iterator<Item = &Entity> {
+        self.inner().values()
+    }
+
+    /// Returns an iterator visiting all component/entity pairs in arbitrary order, with references to the values.
+    pub fn iter(&self) -> impl Iterator<Item = (&ComponentId, &Entity)> {
+        self.inner().iter()
+    }
+
+    pub(crate) fn component_ids(&self) -> &[ComponentId] {
+        self.inner().indices()
+    }
+}
+
+/// A component that contains the resource of type `T`.
+///
+/// When creating a resource, a [`ResourceComponent`] is inserted on a new entity in the world.
+///
+/// This component comes with a hook that ensures that at most one entity has this component for any given `R`:
+/// adding this component to an entity (or spawning an entity with this component) will despawn any other entity with this component.
+/// Moreover, this component requires both marker components [`IsResource`] and [`Internal`].
+/// The former can be used to quickly iterate over all resources through a query,
+/// while the latter marks the associated entity as internal, ensuring that it won't show up on broad queries such as
+/// `world.query::<Entity>()`.
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Component))]
+#[derive(Component)]
+#[require(IsResource, Internal)]
+#[component(on_add = on_add_hook, on_remove = on_remove_hook)]
+#[repr(transparent)]
+pub struct ResourceComponent<R: Resource>(#[entities] pub R);
```

Comment:
```suggestion
// This has an #[entities] annotation because `R` may be a struct which contains entities
pub struct ResourceComponent<R: Resource>(#[entities] pub R);
```

I did not understand this at first!